### PR TITLE
Provides the Performance Flags Framework

### PR DESCRIPTION
**Pull request by andyHa**
https://github.com/scireum/sirius-biz/pull/796/files#diff-abb657b1eb547f3d898c7fe54f6a7c5e

### CODE REVIEW

File: src/main/java/sirius/biz/analytics/flags/ExecutionFlags.java [hunk 1]
```diff
@@ -37,6 +38,19 @@
...
+    /**
+     * Determines if the last execution of the given <tt>flag</tt> for the given <tt>reference</tt> was within the
+     * given <tt>duration</tt>.
+     *
+     * @param reference the entity for which the execution flag is to be determined
+     * @param flag      the flag or type of execution to read
+     * @param duration  the reference duration
+     * @return <tt>true</tt> was stored (set) within the given duration, <tt>false</tt> otherwise
+     */
+    public boolean wasExecuted(BaseEntity<?> reference, String flag, Duration duration) {
```
Comment (sabieber) [inline src/main/java/sirius/biz/analytics/flags/ExecutionFlags.java:50]: why not exactly above the overloaded method?

File: src/main/java/sirius/biz/analytics/flags/PerformanceData.java [hunk 2]
```diff
@@ -0,0 +1,68 @@
...
+import sirius.db.mixing.Composite;
+import sirius.db.mixing.Mapping;
+import sirius.db.mixing.annotations.Transient;
+
+import java.util.stream.Stream;
+
+/**
+ * Can be embedded into an {@link BaseEntity entity} in order to record / toggle performance flags for it.
+ * <p>
+ * Note that this provides a abstract and database independent part. Use {@link SQLPerformanceData} or
```
Comment (sabieber) [inline src/main/java/sirius/biz/analytics/flags/PerformanceData.java:23]:
```suggestion
 * Note that this provides an abstract and database independent part. Use {@link SQLPerformanceData} or
```

File: src/main/java/sirius/biz/analytics/flags/PerformanceData.java [hunk 3]
```diff
@@ -0,0 +1,68 @@
...
+    protected PerformanceData(BaseEntity<?> owner) {
+        this.owner = owner;
+    }
+
+    /**
+     * Obtains a modifier which can be used to efficiently update the flags for the underlying entity.
+     *
+     * @return a modifier used to change the flags for the underlying entity
+     */
+    public abstract PerformanceFlagModifier modifier();
```
Comment (sabieber) [inline src/main/java/sirius/biz/analytics/flags/PerformanceData.java:45]: modify?

File: src/main/java/sirius/biz/analytics/flags/README.md [hunk 4]
```diff
@@ -7,4 +9,14 @@ schedulers which are only to be invoked once per month.
...
+depending on the database being used.
+
+## Performance Flags
+
+Provides a framework which stores a set of flags per entity. This can be either an SQLEntity (which would
+use [SQLPerformanceData](jdbc/SQLPerformanceData.java)) or a MongoEntity (which would use
+[MongoPerformanceData](mongo/MongoPerformanceData.java)). Flags will be stored in a way optimized for the
+underlying database. We use bit fields for JDBC/SQL and StringList for MongoDB. Most probably a performance
+flag will be toggled when computing metrics e.g. via a **MonthlyMetricComputer**. However, they can also
+be togglend on demand using the resepctive [PerformanceFlagModifier](PerformanceFlagModifier.java).
```
Comment (sabieber) [inline src/main/java/sirius/biz/analytics/flags/README.md:21]:
```suggestion
be toggled on demand using the respective [PerformanceFlagModifier](PerformanceFlagModifier.java).
```

File: src/main/java/sirius/biz/analytics/flags/jdbc/BitwiseConstraint.java [hunk 5]
```diff
@@ -0,0 +1,56 @@
...
+
+import sirius.biz.analytics.flags.PerformanceFlag;
+import sirius.biz.analytics.flags.PerformanceFlagged;
+import sirius.db.jdbc.SmartQuery;
+import sirius.db.jdbc.constraints.SQLConstraint;
+
+/**
+ * Provides a constraint which performs a bitwise check in order to determine if a performance flag is toggled.
+ */
+class BitwiseConstraint extends SQLConstraint {
```
Comment (sabieber) [inline src/main/java/sirius/biz/analytics/flags/jdbc/BitwiseConstraint.java:19]: this sounds pretty generic and has performance flag specific implementation. maybe a more specific name is possible? Or a more generic implementation in sirius-db?

File: src/main/java/sirius/biz/analytics/flags/PerformanceFlag.java [hunk 6]
```diff
@@ -17,6 +21,11 @@
 import java.util.Optional;
 import java.util.stream.Stream;
 
+/**
+ * Represents a performance flag which can be toggled for an entity.
+ * <p>
+ * As we use an optimized represnetation, each performance flag must be know ahead of time (declared as constant).
```
Comment (sabieber) [inline src/main/java/sirius/biz/analytics/flags/PerformanceFlag.java:27]:
```suggestion
 * As we use an optimized representation, each performance flag must be know ahead of time (declared as constant).
```

File: src/main/java/sirius/biz/analytics/flags/PerformanceFlag.java [hunk 7]
```diff
@@ -17,6 +21,11 @@
 import java.util.Optional;
 import java.util.stream.Stream;
 
+/**
+ * Represents a performance flag which can be toggled for an entity.
```
Comment (sabieber) [inline src/main/java/sirius/biz/analytics/flags/PerformanceFlag.java:25]: this isnt really descriptive in my eyes

File: src/main/java/sirius/biz/analytics/flags/jdbc/PerformanceFlagConstraint.java [hunk 8]
```diff
@@ -0,0 +1,56 @@
...
+
+    @Override
+    public void appendSQL(SmartQuery.Compiler compiler) {
+        String columnName =
+                compiler.translateColumnName(PerformanceFlagged.PERFORMANCE_DATA.inner(SQLPerformanceData.FLAGS));
+        if (expectedState) {
+            compiler.getWHEREBuilder().append(columnName).append(" & ? <> 0");
+        } else {
+            compiler.getWHEREBuilder().append(columnName).append(" & ? = 0");
+        }
```
Comment (qw3ry) [inline src/main/java/sirius/biz/analytics/flags/jdbc/PerformanceFlagConstraint.java:43]: Hat ein bisschen gedauert, bis ich geblickt hab, dass das bitwise operations sind... vllt n comment? Oder den column name auch als parameter in die query?

File: src/main/java/sirius/biz/analytics/flags/README.md (hunk 4 above)
Comment (idlira) [inline src/main/java/sirius/biz/analytics/flags/README.md]:
```suggestion
be toggled on demand using the respective [PerformanceFlagModifier](PerformanceFlagModifier.java).
```

**Pull request merged**